arangod/Aql/OutputAqlItemRow.cpp

```diff
@@ -0,0 +1,115 @@
+/// @author Jan Christoph Uhde
+////////////////////////////////////////////////////////////////////////////////
+
+#include "OutputAqlItemRow.h"
+
+#include "Aql/AqlItemBlock.h"
+#include "Aql/AqlValue.h"
+#include "Aql/ExecutorInfos.h"
+#include "Aql/InputAqlItemRow.h"
+
+using namespace arangodb;
+using namespace arangodb::aql;
+
+OutputAqlItemRow::OutputAqlItemRow(
+    std::unique_ptr<OutputAqlItemBlockShell> blockShell)
+    : _blockShell(std::move(blockShell)),
+      _baseIndex(0),
+      _inputRowCopied(false),
+      _lastSourceRow{CreateInvalidInputRowHint{}},
+      _numValuesWritten(0) {
+  TRI_ASSERT(_blockShell != nullptr);
+}
+
+void OutputAqlItemRow::setValue(RegisterId registerId,
+                                InputAqlItemRow const& sourceRow,
+                                AqlValue const& value) {
+  if (!isOutputRegister(registerId)) {
+    TRI_ASSERT(false);
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_WROTE_IN_WRONG_REGISTER);
+  }
+  // This is already implicitly asserted by isOutputRegister:
+  TRI_ASSERT(registerId < getNrRegisters());
+  if (_numValuesWritten >= numRegistersToWrite()) {
+    TRI_ASSERT(false);
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_WROTE_TOO_MANY_OUTPUT_REGISTERS);
+  }
+  if (!block().getValueReference(_baseIndex, registerId).isNone()) {
+    TRI_ASSERT(false);
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_WROTE_OUTPUT_REGISTER_TWICE);
+  }
+
+  block().emplaceValue(_baseIndex, registerId, value);
+  _numValuesWritten++;
+  // allValuesWritten() must be called only *after* _numValuesWritten was
+  // increased.
+  if (allValuesWritten()) {
+    copyRow(sourceRow);
+  }
+}
+
+void OutputAqlItemRow::copyRow(InputAqlItemRow const& sourceRow) {
+  TRI_ASSERT(sourceRow.isInitialized());
+  // While violating the following asserted states would do no harm, the
+  // implementation as planned should only copy a row after all values have been
+  // set, and copyRow should only be called once.
+  TRI_ASSERT(!_inputRowCopied);
+  TRI_ASSERT(allValuesWritten());
+  if (_inputRowCopied) {
+    return;
+  }
+
+  // Note that _lastSourceRow is invalid right after construction. However, when
+  // _baseIndex > 0, then we must have seen one row already.
+  TRI_ASSERT(_baseIndex == 0 || _lastSourceRow.isInitialized());
+  bool mustClone = _baseIndex == 0 || _lastSourceRow != sourceRow;
+
+  for (auto itemId : _blockShell->registersToKeep()) {
+    // copy entries to keep
+    //_block->emplaceValue(_baseIndex, itemId, sourceRow.getValue(itemId));
+
+    if (mustClone) {
+      auto const &value = sourceRow.getValue(itemId);
+      if (!value.isEmpty()) {
+        AqlValue clonedValue = value.clone();
+        AqlValueGuard guard(clonedValue, true);
+
+        TRI_IF_FAILURE("OutputAqlItemRow::copyRow") {
+          THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
+        }
+
+        block().setValue(_baseIndex, itemId, clonedValue);
+        guard.steal();
+      }
+    } else {
+      TRI_ASSERT(_baseIndex > 0);
```

Comment:
this has to be moved out of the for loop.
The for is only required during clones, this function references all in one go.